zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -1165,20 +1177,92 @@ private void reconnect() {
     }
   }
 
+  private class SyncContext {
+    private CountDownLatch _latch;
+    private AtomicInteger _rc;
+
+    public SyncContext(CountDownLatch latch) {
+      _latch = latch;
+    }
+
+    AtomicInteger getRc() {
+      return _rc;
+    }
+
+    void setRc(AtomicInteger rc) {
+      _rc = rc;
+    }
+
+    CountDownLatch getLatch() {
+      return _latch;
+    }
+  }
+
+  private boolean retrySync(String sessionId) throws ZkInterruptedException {
+    if (!_syncOnNewSession) {
+      return true;
+    }
+    while (true) {
+      CountDownLatch latch = new CountDownLatch(1);
+      SyncContext ctx = new SyncContext(latch);
+
+      final ZkConnection zkConnection = (ZkConnection) getConnection();
+      zkConnection.getZookeeper().sync(_syncPath, new AsyncCallback.VoidCallback() {
+        @Override
+        public void processResult(int rt, String s, Object ctx) {
+          LOG.info("sycnOnNewSession with sessionID {} async return code: {}", sessionId, rt);
+          SyncContext synCtx = ((SyncContext) ctx);
+          synCtx.setRc(new AtomicInteger(rt));
+          synCtx.getLatch().countDown();
+        }
+      }, ctx);
+
+      try {
+        latch.await();
+      } catch (InterruptedException e) {
+        LOG.info("retrySync latch waiting got interrrupted with sessionId {} ", sessionId);
+        throw new ZkInterruptedException(e);
+      }
+
+      KeeperException.Code code = KeeperException.Code.get(ctx.getRc().get());
+      if (code == OK) {
+        LOG.info("sycnOnNewSession with sessionID {} async return code: {} and proceeds", sessionId,
+            code);
+        break;
+      }
+      if (code == CONNECTIONLOSS || code == SESSIONMOVED) {
+        LOG.info("sycnOnNewSession with sessionID {} async return code: {} and retry", sessionId,
+            code);
+        continue;
+      }
+      // Not retryable, including session expiration; but having error. Log the error and return
+      LOG.error(
+          "sycnOnNewSession with sessionID {} async return code: {} and not retryable, stop calling handleNewSession",
+          sessionId, code);
+      return false;
+    }
+    return true;
+  }
+
   private void fireNewSessionEvents() {
     // only managing zkclient fire handleNewSession event
     if (!isManagingZkConnection()) {
       return;
     }
     final String sessionId = getHexSessionId();
     for (final IZkStateListener stateListener : _stateListener) {
-      _eventThread.send(new ZkEventThread.ZkEvent("New session event sent to " + stateListener, sessionId) {
+      _eventThread
+          .send(new ZkEventThread.ZkEvent("New session event sent to " + stateListener, sessionId) {
 
-        @Override
-        public void run() throws Exception {
-          stateListener.handleNewSession(sessionId);
-        }
-      });
+            @Override
+            public void run() throws Exception {
+              boolean proceed = retrySync(sessionId);
```

Comment:
Do we need to do this for every single state listener?
We shall only do it once, right?

Reply:
+1. This is not necessary doing sync for each of listener. Although once the first one retry complete rest will not cost much time, this is redundant logic. We should avoid it.

Reply:
This is a very good catch. Will change it to be per session, only once.